stdlib/public/core/String.swift

```diff
@@ -981,115 +981,119 @@ extension String {
-  var icuOutputBuffer = icuOutputBuffer
-
-  var index = String.Index(_encodedOffset: 0)
-  let cachedEndIndex = String.Index(_encodedOffset: sourceBuffer.count)
-
-  var hasBufferOwnership = false
-
-  defer {
-    if hasBufferOwnership {
-      outputBuffer.deallocate()
-      icuInputBuffer.deallocate()
-      icuOutputBuffer.deallocate()
+  internal func _withNFCCodeUnits(_ f: (UInt8) throws -> Void) rethrows {
+    // Fast path: If we're already NFC (or ASCII), then we don't need to do
+    // anything at all.
+    if _guts.isNFC {
+      try _forEachUTF8(f)
+      return
     }
-  }
 
-  while index < cachedEndIndex {
-    let result = _fastNormalize(
-      readIndex: index,
-      sourceBuffer: sourceBuffer,
-      outputBuffer: &outputBuffer,
-      icuInputBuffer: &icuInputBuffer,
-      icuOutputBuffer: &icuOutputBuffer
-    )
-    for i in 0..<result.amountFilled {
-      try f(outputBuffer[i])
+    var isNFCQC = true
+
+    if _guts.isFastUTF8 {
+      _guts.withFastUTF8 { utf8 in
+        var position = 0
+        var prevCCC: UInt16 = 0
+
+        while position < utf8.count {
+          // If our first byte is less than 0xCC, then it means we're under the
+          // 0x300 scalar value and everything up to 0x300 is NFC already.
+          if utf8[position] < 0xCC {
+            // If our first byte is less than 0xC0, then it means it is ASCII
+            // and only takes up a single byte.
+            if utf8[position] < 0xC0 {
+              position &+= 1
+            } else {
+              // Otherwise, this is a 2 byte < 0x300 sequence.
+              position &+= 2
+            }
+            // ASCII always has ccc of 0.
+            prevCCC = 0
+
+            continue
+          }
+
+          let (scalar, len) = _decodeScalar(utf8, startingAt: position)
+
+          if !_isScalarNFCQC(scalar, &prevCCC) {
+            isNFCQC = false
+            return
+          }
+
+          position &+= len
+        }
+      }
+
+      // Because we have access to the fastUTF8, we can go through that instead
+      // of accessing the UTF8 view on String.
+      if isNFCQC {
+        try _guts.withFastUTF8 {
+          for byte in $0 {
+            try f(byte)
+          }
+        }
+
+        return
+      }
+    } else {
+      var prevCCC: UInt16 = 0
+
+      for scalar in String(_guts).unicodeScalars {
+        if !_isScalarNFCQC(scalar, &prevCCC) {
+          isNFCQC = false
+          break
+        }
+      }
+
+      if isNFCQC {
+        for byte in String(_guts).utf8 {
+          try f(byte)
+        }
+
+        return
+      }
     }
-    _internalInvariant(result.nextReadPosition != index)
-    index = result.nextReadPosition
-    if result.allocatedBuffers {
-      _internalInvariant(!hasBufferOwnership)
-      hasBufferOwnership = true
+
+    for scalar in String(_guts).nfc {
+      try scalar.withUTF8CodeUnits {
+        for byte in $0 {
+          try f(byte)
+        }
+      }
```

Comment:
If we want to make a public `String.nfc` var, then all this code should probably be in its iterator's `next()` method.